aws/credentials/processcreds/provider.go

```diff
@@ -0,0 +1,353 @@
+
+Another way to use the `credential_process` is by using
+`credentials.NewCredentials()` and providing a process to be executed to
+retrieve credentials:
+
+	// Create credentials using the ProcessProvider.
+	creds := credentials.NewCredentials(&processcreds.ProcessProvider{
+		Process: "/path/to/process",
+	})
+
+	// Create service client value configured for credentials.
+	svc := s3.New(sess, &aws.Config{Credentials: creds})
+
+A similar way to use the `credential_process` is by using the `ProcessProvider`
+itself.
+
+	// Create the ProcessProvider.
+	p := &processcreds.ProcessProvider{
+		Process: "/path/to/process",
+	})
+	creds := p.NewCredentials()
+*/
+package processcreds
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+)
+
+// ProviderName is the name this credentials provider will label any returned
+// credentials Value with.
+const ProviderName = `ProcessProvider`
+
+const (
+	// ErrCodeProcessProviderParse error parsing process output
+	ErrCodeProcessProviderParse = "ProcessProviderParseError"
+
+	// ErrCodeProcessProviderVersion version error in output
+	ErrCodeProcessProviderVersion = "ProcessProviderVersionError"
+
+	// ErrCodeProcessProviderRequired required attribute missing in output
+	ErrCodeProcessProviderRequired = "ProcessProviderRequiredError"
+
+	// ErrCodeProcessProviderExecution execution of process failed
+	ErrCodeProcessProviderExecution = "ProcessProviderExecutionError"
+
+	// DefaultDuration is the default amount of time in minutes that the credentials
+	// will be valid for.
+	DefaultDuration = time.Duration(15) * time.Minute
+
+	// DefaultInitialBufSize size for initial buffer.
+	DefaultInitialBufSize = 200
+
+	// DefaultMaxBufSize limits memory usage from growing to an enormous
+	// amount due to a faulty process.
+	DefaultMaxBufSize = 512
+
+	// DefaultTimeout limits the time a process can run, in milliseconds.
+	DefaultTimeout = 500
+)
+
+// ProcessProvider satisfies the credentials.Provider interface, and is a client to
+// retrieve credentials from a process.
+type ProcessProvider struct {
+	staticCreds bool
+	credentials.Expiry
+
+	// Expiry duration of the credentials. Defaults to 15 minutes if not set.
+	Duration time.Duration
+
+	// ExpiryWindow will allow the credentials to trigger refreshing prior to
+	// the credentials actually expiring. This is beneficial so race conditions
+	// with expiring credentials do not cause request to fail unexpectedly
+	// due to ExpiredTokenException exceptions.
+	//
+	// So a ExpiryWindow of 10s would cause calls to IsExpired() to return true
+	// 10 seconds before the credentials are actually expired.
+	//
+	// If ExpiryWindow is 0 or less it will be ignored.
+	ExpiryWindow time.Duration
+
+	// A string representing an os command that should return a JSON with
+	// credential information.
+	Process string
```

Comment:
How about instead of `Process` being a string, but a `exec.Cmd` or some sort of `Command` type? This will allow for all the OS specific logic to be very contained in the `Command` portion.